https://issues.redhat.com/browse/ACM-23917 MCOA (#8364)
open-cluster-management/rhacm-docs: observability/obs_mcoa_add_custom_metrics.adoc

[#add-custom-metrics-mcoa]
= Adding custom metrics for the multicluster observability add-on

Add your own custom metrics to be collected from your managed clusters by configuring the `ScrapeConfig` resource. The new `ScrapeConfig` resource is deployed to the managed clusters that you specify in the placements of the `ClusterManagementAddOn` resource. The Prometheus operator on each managed cluster adds the new `ScrapeConfig` resource to the `PrometheusAgent` resource.

*Requierd access:* Cluster administrator

.Prerequisites

- You have installed and enabled the multicluster observability add-on. For more information, see xref:../observability/obs_enable_mcoa.adoc#enable-mcoa[Enabling the multicluster observability add-on].

Complete the following steps to add custom metrics for the multicluster observability add-on:

. Create a new `ScrapeConfig` resource in the `open-cluster-management-observability` namespace that includes values for the required parameters, `jobName`, `metricsPath`, and `params`.

> . Add the appropriate label for the `app.kubernetes.io/component` parameter to specify whether the metrics are for platform monitoring or user workload monitoring. Use one of the following label values, `platform-metrics-collector` or `user-workload-metrics-collector`.

Reviewer: Maybe add the when `platform-metrics-collector` is set, MCOA will automatically set the `scrapeClass` and the `targets` to enable federation from the in-cluster platform prometheus. While when `user-workload-metrics-collector`, they will be automatically set to target the in-cluster user-workload monitoring. But this can be overridden by the user in the latter case.

Author: Is there an example that we can share with the user? Maybe a YAML snippet ? Also what does "in-cluster" mean? Are you referring to the hub cluster?

Reviewer: "in-cluster platform prometheus" refers to the ocp monitoring prometheus, that operates for this single cluster.

Here is an example:

```yaml
apiVersion: monitoring.rhobs/v1alpha1
kind: ScrapeConfig
metadata:
  name: platform-metrics-alerts
  namespace: open-cluster-management-agent-addon
spec:
  jobName: alerts
  metricRelabelings:
    - action: labeldrop
      regex: managed_cluster|id
  metricsPath: /federate
  params:
    'match[]':
      - '{__name__="ALERTS"}'
  scheme: HTTPS
  scrapeClass: ocp-monitoring
  staticConfigs:
    - targets:
        - 'prometheus-k8s.openshift-monitoring.svc:9091'
```

jobName, params and metricsPath is set by the user while the other fields are enforced by the operator. But for metricRelabelings, the user can still add his own relabellings on top of ours.

Reviewer: And for UWL scrapeconfigs, we don't enforce scrapeClass and  staticConfigs. We only set a default that targets the ocp uwl monitoring prometheus. But the user can override that.

. Add the `ScrapeConfig` resource reference to the placements of the `ClusterManagementAddOn` resource where you want the resource to be deployed.

+

> *Note:* Ensure that you reference the `ScrapeConfig` resource only after each federated metric is deployed on a given managed cluster.

Reviewer: I suppose you mean:
Ensure that you reference the `ScrapeConfig` resource only after it has been created. Otherwise, the addon status will become `Deploying` as the resource is not found.

+
Your `ScrapeConfig` resource might resemble the following YAML file:
//please help with completing this yaml example
+
[source,yaml]
----
apiVersion: monitoring.rhobs/v1alpha1
kind: ScrapeConfig
metadata:
  name: add-custom-metrics
  namespace: open-cluster-management-observability
  labels:

>     - app.kubernetes.io/component: <platform-metrics-collector>

Reviewer: Let's remove the <>, no?

[#add-custom-metrics-mcoa]
= Adding custom metrics for the multicluster observability add-on

> Add your own custom metrics to be collected from your managed clusters by configuring the `ScrapeConfig` resource. The new `ScrapeConfig` resource is deployed to the managed clusters that you specify in the placements of the `ClusterManagementAddOn` resource. The Prometheus operator on each managed cluster adds the new `ScrapeConfig` resource to the `PrometheusAgent` resource.

Reviewer: ```suggestion
Add your own custom metrics to be collected from your managed clusters by configuring the `ScrapeConfig` resource. The new `ScrapeConfig` must be added to the `Placement` configurations of the `ClusterManagementAddOn` resource for deploying on the corresponding managed clusters. The Prometheus operator on each managed cluster adds the new `ScrapeConfig` resource to the `PrometheusAgent` resource.
```

[#add-custom-metrics-mcoa]
= Adding custom metrics for the multicluster observability add-on

> Add your own custom metrics to be collected from your managed clusters by configuring the `ScrapeConfig` resource. The new `ScrapeConfig` must be added to the `Placement` configurations of the `ClusterManagementAddOn` resource for deploying on the corresponding managed clusters. The Prometheus operator on each managed cluster adds the new `ScrapeConfig` resource to the `PrometheusAgent` resource.

Reviewer 2: Just thinking ahead, if we say something like "new", it will be outdated by the next version, unless we remember to come back to this and change it. Should we just leave it out now? Let me know what you think.

Author: Thats a great point. I'll remove new

*Requierd access:* Cluster administrator

.Prerequisites

- You have installed and enabled the multicluster observability add-on. For more information, see xref:../observability/obs_enable_mcoa.adoc#enable-mcoa[Enabling the multicluster observability add-on].

Complete the following steps to add custom metrics for the multicluster observability add-on:

. Create a new `ScrapeConfig` resource in the `open-cluster-management-observability` namespace that includes values for the required parameters, `jobName`, `metricsPath`, and `params`.

. Add the appropriate label for the `app.kubernetes.io/component` parameter to specify whether the metrics are for platform monitoring or user workload monitoring. Use one of the following label values, `platform-metrics-collector` or `user-workload-metrics-collector`.

+

> *Note:* When you use the `platform-metrics-collector` label, the multicluster observability add-on automatically set the `scrapeClass` and `targets` parameters to enable federation from the platform Prometheus of your {ocp-short} managed cluster. You can override the `scrapeClass` and `targets` parameters by adding the value that you need.

Reviewer 2: ```suggestion
*Note:* When you use the `platform-metrics-collector` label, the multicluster observability add-on automatically sets the `scrapeClass` and `targets` parameters to enable federation from the platform Prometheus of your {ocp-short} managed cluster. You can override the `scrapeClass` and `targets` parameters by adding the value that you need.
```
